### DISCUSSION TITLE
First feedback from a quick try

### DISCUSSION THREAD

**Discussion by kvark** [pest-parser/pest#637] [General]
Some things that I found sub-ideal:
  1. lack of #446
  2. path in `#[grammar = "ident.pest"]` appears to be neither from root or local to the file, but instead hardcoded to "$root/src"
  3. in the example of https://github.com/pest-parser/pest#pairs-api the `Rule` comes out of nowhere

**Comment by dragostis** [comment]
@kvark, thanks for the feedback. For no. 2, I haven't really kept track of proc macro developments as of late, but I'm not completely sure whether there's a way to get the path of the file where the proc macro was invoked so you can do relative inclusion of the file.

Once we have this possibility, ideally we would try both relative and cargo root paths:

  * if only relative path is found, use that
  * if only cargo root path is found, show a warning that cargo root paths have been deprecated
  * if both are found either show a warning, or show an error; the error would be a breaking change, but it should not be so common

**Comment by kvark** [comment]
The standard `include!` macro and friends use the local path, perhaps you could follow the same logic.

**Comment by CAD97** [comment]
Yeah, that'd be great but:

A) it's not possible currently. There is no way for a procedural macro to open a path relative to the file that is being expanded for. You can get the manifest directory (and whatever the working directory happens to be; that's not specified and even changed at some point, breaking some macros) and that's it. If I'm wrong, I'd love to be corrected.

B) changing the path base is a breaking change, so the current behavior needs to at the very least be supported (potentially with a deprecation warning) once we _can_ act like `include!`.
